mat64/shadow.go

```diff
@@ -223,3 +225,55 @@ func rectanglesOverlap(off, aCols, bCols, stride int) bool {
 	// b strictly wraps and so must overlap with a.
 	return true
 }
+
+// trianglesOverlap returns whether the strided triangles a and b overlap
+// when b is offset by off elements after a but has at least one element before
+// the end of a. a and b have aSize and bSize respectively.
+func trianglesOverlap(off, aSize, bSize, stride int, aUpper, bUpper bool) bool {
+	if !rectanglesOverlap(off, aSize, bSize, stride) {
+		// Fast return if bounding rectangles do not overlap.
+		return false
+	}
+
+	// Find location of b relative to a.
+	rowOffset := off / stride
+	colOffset := off % stride
+	if (off+bSize)%stride < colOffset {
+		// We have wrapped, so readjust offsets.
+		rowOffset++
+		colOffset -= stride
+	}
+
+	if aUpper {
+		// Check whether the upper left of b
+		// is in the triangle of a
+		if rowOffset >= 0 && rowOffset <= colOffset {
+			return true
+		}
+		// Check whether the upper right of b
+		// is in the triangle of a.
+		return bUpper && rowOffset < colOffset+bSize
+	}
+
+	// Check whether the upper left of b
+	// is in the triangle of a
+	if colOffset >= 0 && rowOffset >= colOffset {
+		return true
+	}
+	if !bUpper {
```

Comment:
If you tested here `if bUpper` instead, there could be one indentation level less. Just a cosmetic detail.

I'm wondering if the following arrangement (and similar) is being tested somewhere:
```
a
aa
a.a b
a..abb
a...b.b
aaaab..b
    bbbbb
```